asllib/doc/asl.spec

```
@@ -1697,7 +1742,340 @@ typing relation annotate_expr(tenv: static_envs, e: expr) -> (t: ty, new_e: expr
+          assoc_opt(fields, field_name) = some(t);
+          --
+          (t, E_GetField(e2, field_name), ses1);
+        }
+        case error {
+          assoc_opt(fields, field_name) = None;
+          --
+          TypeError(TE_BF);
+        }
+      }
+      case collection {
+        L = T_Collection;
+        case okay {
+          assoc_opt(fields, field_name) = Some(t);
+          --
+          (t, E_GetCollectionFields(base, make_list(field_name)), ses1)
+          { math_layout = [_] };
+        }
+        case error {
+          assoc_opt(fields, field_name) = None;
+          --
+          TypeError(TE_BF);
+        }
+      }
+    }
+
+    case bitfield {
+      T_Bits(_, bitfields) := t_e2;
+      case simple {
+        find_bitfield_opt(bitfields, field_name) -> some(BitField_Simple(_, slices));
+        e3 := E_Slice(e2, slices);
+        annotate_expr(tenv, e3) -> (t, new_e, ses);
+        --
+        (t, new_e, ses);
+      }
+      case nested {
+        find_bitfield_opt(bitfields, field_name) -> some(BitField_Nested(_, slices, bitfields'))
+        { math_layout = [_] };
+        e3 := E_Slice(e2, slices);
+        annotate_expr(tenv, e3) -> (t, new_e, ses_new);
+        T_Bits(width, _) := t_e4;
+        t := T_Bits(width, bitfields');
+        --
+        (t, new_e, ses_new);
+      }
+      case typed {
+        find_bitfield_opt(bitfields, field_name) -> some(BitField_Type(_, slices, t))
+        { math_layout = [_] };
+        e3 := E_Slice(e2, slices);
+        annotate_expr(tenv, e3) -> (t, new_e, ses_new);
+        checked_typesat(tenv, t_e4, t) -> True;
+        --
+        (t, new_e, ses_new);
+      }
+      case error {
+        find_bitfield_opt(bitfields, field_name) -> None;
+        --
+        TypeError(TE_BF);
+      }
+    }
+
+    case tuple_item {
+      t_e2 =: T_Tuple(tys);
+      concat_strings(item, num) := field_name;
+      num in Lang(int_lit_regex);
+      decimal_to_lit(num) = INT_LIT(index);
+      te_check(zero <= index && index <= list_len(tys), TE_BTI) -> True;
+      t := tys[index];
+      new_e := E_GetItem(e2, index);
+      --
+      (t, new_e, ses1);
+    }
+
+    case error {
+      ast_label(t_e2) not_in make_set(T_Record, T_Exception, T_Collection, T_Bits, T_Tuple)
+      { math_layout = (_, [_]) };
+      --
+      TypeError(TE_UT);
+    }
+  }
+
+  // Missing specification:
+  case EGetFields {
+    e =: E_GetFields(e1, fields);
+    annotate_expr(tenv, e_base) -> (t_base_annot, e_base_annot, ses_base);
+
+    case bits {
+      make_anonymous(tenv, t_base_annot) -> T_Bits(_, bitfields);
+      INDEX(i, fields: find_bitfields_slices(fields[i], bitfields) -> slices[i]);
+      e_slice := E_Slice(e_base, slices);
+      annotate_expr(tenv, e_slice) -> (t, new_e, ses);
+      --
+      (t, new_e, ses);
+    }
+    case record {
+      make_anonymous(tenv, t_base_annot) -> T_Record(base_fields);
+      INDEX(i, base_fields: get_bitfield_width(tenv, f, tfields) -> e_width);
```

Should this be:
```suggestion
      INDEX(i, base_fields: get_bitfield_width(tenv, f, tfields) -> e_width[i]);
```
?
(similarly below)